drivers/rf-transceiver/hmc6300/hmc6300.c

```diff
@@ -0,0 +1,439 @@
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <errno.h>
+#include <stdlib.h>
+#include "hmc6300.h"
+#include "no_os_delay.h"
+
+const uint8_t hmc6300_default_regmap[] = {
+	0x00, 0x4a, 0xf6, 0xf6, 0x00, 0xbf, 0x6c, 0x0f,
+	0x8f, 0x00, 0x53, 0x03, 0x00, 0x00, 0x00, 0x00,
+	0x36, 0xbb, 0x46, 0x02, 0x22, 0x12, 0x00, 0x62
+};
+
+static int _hmc6300_generate_vco(struct hmc6300_dev *d, uint64_t f_start,
+				 uint64_t f_step, uint8_t fbdiv_start, uint8_t entries)
+{
+	int ret, e;
+
+	d->vco.freqs = (uint64_t *) calloc(entries, sizeof(uint64_t));
+	if (!d->vco.freqs)
+		return -ENOMEM;
+
+	d->vco.fbdiv = (uint8_t *) calloc(entries, sizeof(uint8_t));
+	if (!d->vco.fbdiv) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	for(e = 0; e < entries; e++) {
+		d->vco.freqs[e] = f_start + e * f_step;
+		d->vco.fbdiv[e] = fbdiv_start + e;
+	}
+
+	d->vco.entries = entries;
+
+	return 0;
+error:
+	free(d->vco.freqs);
+	return ret;
+}
+
+int hmc6300_init(struct hmc6300_dev **dev, struct hmc6300_init_param *init)
+{
+	struct hmc6300_dev *d;
+	int ret;
+
+	if (!dev || !init)
+		return -EINVAL;
+
+	d = (struct hmc6300_dev *) calloc(1, sizeof(*d));
+	if (!d)
+		return -ENOMEM;
+
+	switch(init->ref_clk) {
+	case HMC6300_REF_CLK_71p42857MHz:
+		ret = _hmc6300_generate_vco(d, 56500000000, 250000000, 0x22, 31);
+		break;
+	case HMC6300_REF_CLK_75MHz:
+		ret = _hmc6300_generate_vco(d, 55125000000, 262500000, 0x12, 43);
+		break;
+	case HMC6300_REF_CLK_142p8571MHz:
+		ret = _hmc6300_generate_vco(d, 56500000000, 500000000, 0x11, 16);
+		break;
+	case HMC6300_REF_CLK_154p2857MHz:
+		ret = _hmc6300_generate_vco(d, 57240000000, 540000000, 0x0a, 17);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_get(&d->en, &init->en);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_get(&d->clk, &init->clk);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_get(&d->data, &init->data);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_get(&d->scanout, &init->scanout);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_direction_output(d->en, NO_OS_GPIO_HIGH);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_direction_output(d->clk, NO_OS_GPIO_LOW);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_direction_output(d->data, NO_OS_GPIO_LOW);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_direction_input(d->scanout);
+	if (ret)
+		goto error;
+
+	ret = hmc6300_write_regmap(d, hmc6300_default_regmap);
+	if (ret)
+		goto error;
+
+	*dev = d;
+	return 0;
+error:
+	free(d);
+	return ret;
+}
+
+int hmc6300_remove(struct hmc6300_dev *dev)
+{
+	free(dev->vco.freqs);
+	free(dev->vco.fbdiv);
+	no_os_gpio_remove(dev->en);
```

Comment:
Hmm, I was also wondering whether I should do it or not. To me, if a remove() fails, so be it, but we need to remove the others too, otherwise we end up with memory leaks.